text/0000-macro-expansion-for-macro-input.md

```diff
@@ -0,0 +1,214 @@
+
+If a marked `#` token is part of an inner-attribute `#![...]` call, the situation is similar: the parser can forward-parse the macro path and token arguments, and with a little work can forward-parse the body.
+
+When the parser encounters a marked `!` token, it needs to forward-parse the token arguments, but also needs to _backtrack_ to parse the macro path. In a structured area of the grammar (such as in an attribute macro body or a structured decl macro) this would be fine, since we would already be parsing an expression or item and hence have the path ready. In an _unstructured_ area we would actually have to backtrack within the token stream and 'reverse parse' a path: is this an issue?
+
+## Delayed resolution of unresolved macros
+
+The existing expansion loop adds any currently unresolved macros to a _resolution_ queue. When re-parsing macro output, if any newly defined macros would allow those unresolved macros to be resolved, they get added to the current expansion queue. If there are unresolved macros but no macros to expand, the compiler reports the unresolvable definition.
+
+The new expansion order described [above](#macro-expansion-and-marking) is designed to expand all marked macros as much as possible before trying to expand unmarked ones. We know that marked macros are always in token position, so expansion-eligible unmarked macros are the only way to introduce new macro definitions.
+
+In the new order, we still accumulate unresolved macros (marked and unmarked), and we still remove them from the resolution queue to the relevant expansion queue whenever they get defined. The only difference is an extra error case, where a resolved unmarked macro has an unresolved marked macro in its input, and there are no unmarked macros to expand. In this case, the resolution queue still contains the unresolved marked macro, and so the compiler again reports the unresolvable definition.
+
+# Drawbacks
+
+This proposal:
+
+* Relies on proc macro authors doing macro expansion. This might partition the macro ecosystem into expansion-ignoring (where input macro calls are essentially forbidden for any part of the input that needs to be inspected) and expansion-handling (where they work fine _as long as_ the proc macro author has used the expansion API correctly).
+
+* Leads to frustrating corner-cases involving macro paths. For instance, consider the following:
+
+    ```rust
+    macro baz!(...);
+    foo! {
+        mod b {
+            super::baz!();
+        }
+    }
+    ```
+
+    The caller of `foo!` probably imagines that `baz!` will be expanded within `mod b`, and so prepends the call with `super`. However, if `foo!` naively marks the call to `super::baz!`, then the path will fail to resolve because macro paths are resolved relative to the location of the call. Handling this would require the macro implementer to track the path offset of its expansion, which is doable but adds complexity.
```

Comment:
Could the resolver look into module structure within the `TokenStream`? Or would this break backwards compatibility? 

We should also note that this only applies to procedural bang-macros, because with attributes, the resolver already has the AST available, including any possible macro structure, so unless they change the latter, they can safely ignore it.

Reply:
Letting the compiler look in to the token stream beyond identifying the macro call sounds possibly risky - there are a lot more parsing corner cases (what if my marked macro _just so happens_ to be near a bunch of unstructured tokens that are structured like a `mod`?). If we were proposing marking entire items or expressions for the compiler to interpret I'd be more comfortable with that.

Reply:
Sounds like a plan. Mark a complete TokenStream to have all macros within expanded? That would at the very least greatly simplify the usage for macro authors, at the cost of some complexity for expansion.